test/hack/cleanup/main.go

```diff
@@ -51,59 +61,171 @@ func main() {
+		&stack{cloudFormationClient: cloudFormationClient, cloudWatchClient: cloudWatchClient, expirationTime: expirationTime, logger: logger},
+		&launchtemplate{ec2Client: ec2Client, cloudWatchClient: cloudWatchClient, expirationTime: expirationTime, logger: logger},
+		&oidc{iamClient: iamClient, cloudWatchClient: cloudWatchClient, expirationTime: expirationTime, logger: logger},
+	}
 
-	// Terminate any old instances that were provisioned by Karpenter as part of testing
-	// We execute these in serial since we will most likely get rate limited if we try to delete these too aggressively
-	ids := getOldInstances(ctx, ec2Client, expirationTime)
-	logger.With("ids", ids, "count", len(ids)).Infof("discovered test instances to delete")
+	workqueue.ParallelizeUntil(ctx, 10, len(resources), func(i int) {
+		resources[i].cleanupResources(ctx)
+	})
+}
+
+type instance struct {
+	ec2Client        *ec2.Client
+	cloudWatchClient *cloudwatch.Client
+	expirationTime   time.Time
+	logger           *zap.SugaredLogger
+}
+
+// Terminate any old instances that were provisioned by Karpenter as part of testing
+// We execute these in serial since we will most likely get rate limited if we try to delete these too aggressively
+func (i *instance) cleanupResources(ctx context.Context) {
+	ids, err := i.getOldResources(ctx)
+	if err != nil {
+		i.logger.With("error", err).Error("getting instances")
+	}
+	i.logger.With("ids", ids, "count", len(ids)).Infof("discovered test instances to delete")
 	if len(ids) > 0 {
-		if _, err := ec2Client.TerminateInstances(ctx, &ec2.TerminateInstancesInput{
+		if _, err := i.ec2Client.TerminateInstances(ctx, &ec2.TerminateInstancesInput{
 			InstanceIds: ids,
 		}); err != nil {
-			logger.With("ids", ids, "count", len(ids)).Errorf("terminating test instances, %v", err)
+			i.logger.With("ids", ids, "count", len(ids)).Errorf("terminating test instances, %v", err)
 		} else {
-			logger.With("ids", ids, "count", len(ids)).Infof("terminated test instances")
-			if err = fireMetric(ctx, cloudWatchClient, "InstancesDeleted", float64(len(ids))); err != nil {
-				logger.With("name", "InstancesDeleted").Errorf("firing metric, %v", err)
+			i.logger.With("ids", ids, "count", len(ids)).Infof("terminated test instances")
+			if err = fireMetric(ctx, i.cloudWatchClient, "InstancesDeleted", float64(len(ids))); err != nil {
+				i.logger.With("name", "InstancesDeleted").Errorf("firing metric, %v", err)
 			}
 		}
 	}
+}
 
-	// Terminate any old stacks that were provisioned as part of testing
-	// We execute these in serial since we will most likely get rate limited if we try to delete these too aggressively
-	names := getOldStacks(ctx, cloudFormationClient, expirationTime)
-	logger.With("names", names, "count", len(names)).Infof("discovered test stacks to delete")
+type securitygroup struct {
+	ec2Client        *ec2.Client
+	cloudWatchClient *cloudwatch.Client
+	expirationTime   time.Time
```

Comment:
Same comment on expiration time? Each of these resource structs has this value so is there a good way to standardize this so it's more DRY